src/wp-includes/block-patterns.php

```diff
@@ -319,37 +319,132 @@ function _register_remote_theme_patterns() {
+			}
+
+			register_block_pattern( $pattern_data['slug'], $pattern_data );
+		}
+	}
+}
+add_action( 'init', '_register_theme_block_patterns' );
+
+/**
+ * Gets block pattern data for a specified theme.
+ * Each pattern is defined as a PHP file and defines
+ *  its metadata using plugin-style headers. The minimum required definition is:
  *
- * The output of the PHP source corresponds to the content of the pattern, e.g.:
+ *      /**
+ *       * Title: My Pattern
+ *       * Slug: my-theme/my-pattern
+ *       *
  *
- *     <main><p><?php echo "Hello"; ?></p></main>
+ *  The output of the PHP source corresponds to the content of the pattern, e.g.:
  *
- * If applicable, this will collect from both parent and child theme.
+ *      <main><p><?php echo "Hello"; ?></p></main>
  *
- * Other settable fields include:
+ *  If applicable, this will collect from both parent and child theme.
  *
- *   - Description
- *   - Viewport Width
- *   - Inserter         (yes/no)
- *   - Categories       (comma-separated values)
- *   - Keywords         (comma-separated values)
- *   - Block Types      (comma-separated values)
- *   - Post Types       (comma-separated values)
- *   - Template Types   (comma-separated values)
+ *  Other settable fields include:
  *
- * @since 6.0.0
- * @since 6.1.0 The `postTypes` property was added.
- * @since 6.2.0 The `templateTypes` property was added.
+ *    - Description
+ *    - Viewport Width
+ *    - Inserter         (yes/no)
+ *    - Categories       (comma-separated values)
+ *    - Keywords         (comma-separated values)
+ *    - Block Types      (comma-separated values)
+ *    - Post Types       (comma-separated values)
+ *    - Template Types   (comma-separated values)
+ *
+ * @since 6.4.0
  * @access private
+ *
+ * @param WP_Theme $theme Theme object.
+ * @return array Block pattern data.
  */
-function _register_theme_block_patterns() {
+
+function _wp_get_block_patterns( WP_Theme $theme ) {
+	if ( ! $theme->exists() ) {
+		return array(
+			'version'  => false,
+			'patterns' => array(),
+		);
+	}
+
+	$transient_name = 'wp_theme_patterns_' . $theme->get_stylesheet();
+	$version        = $theme->get( 'Version' );
+	$can_use_cached = ! wp_is_development_mode( 'theme' );
+
+	if ( $can_use_cached ) {
+		$pattern_data = get_transient( $transient_name );
+		if ( is_array( $pattern_data ) && $pattern_data['version'] === $version ) {
+			return $pattern_data;
+		}
+	}
+
+	$pattern_data = array(
+		'version'  => $version,
+		'patterns' => array(),
+	);
+	$dirpath      = $theme->get_stylesheet_directory() . '/patterns/';
+
+	if ( ! file_exists( $dirpath ) ) {
+		if ( $can_use_cached ) {
+			set_transient( $transient_name, $pattern_data );
+		}
+		return $pattern_data;
+	}
+	$files = glob( $dirpath . '*.php' );
+	if ( ! $files ) {
+		if ( $can_use_cached ) {
+			set_transient( $transient_name, $pattern_data );
+		}
+		return $pattern_data;
+	}
```

Comment:
See my other suggestion, we can simplify this a lot by abstracting into `WP_Theme`. With that, this code here would become:
```suggestion
	$can_use_cached = ! wp_is_development_mode( 'theme' );

	if ( $can_use_cached ) {
		$patterns = $theme->get_pattern_cache();
		if ( is_array( $patterns ) ) {
			return $patterns;
		}
	}

	$patterns = array();
	$dirpath  = $theme->get_stylesheet_directory() . '/patterns/';

	if ( ! file_exists( $dirpath ) ) {
		if ( $can_use_cached ) {
			$theme->set_pattern_cache( $patterns );
		}
		return $patterns;
	}
	$files = glob( $dirpath . '*.php' );
	if ( ! $files ) {
		if ( $can_use_cached ) {
			$theme->set_pattern_cache( $patterns );
		}
		return $patterns;
	}
```